"Fatal Unhandled Exception" error for Sequence Construction Expressions

The following code generates the fatal unhandled exception:
```dafny
function method FromAtoB(a: seq<int>): (b: seq<int>) {
    seq(|a|, i => a[i])
}
```
```bash
[ERROR] FATAL UNHANDLED EXCEPTION: System.Diagnostics.Contracts.ContractException: Assertion failed.
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailureImplementation (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0002f] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailure (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x00000] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Diagnostics.Contracts.Contract.ReportFailure (System.Diagnostics.Contracts.ContractFailureKind failureKind, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0003a] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Diagnostics.Contracts.Contract.Assert (System.Boolean condition) [0x00003] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at Microsoft.Boogie.NAryExpr..ctor (Microsoft.Boogie.IToken tok, Microsoft.Boogie.IAppliable fun, System.Collections.Generic.IList`1[T] args, System.Boolean immutable) [0x00043] in <451e59e8cecd4b8a8d901251129e1be1>:0
  at Microsoft.Dafny.Translator.FunctionCall (Microsoft.Boogie.IToken tok, System.String function, Microsoft.Boogie.Type returnType, Microsoft.Boogie.Expr[] args) [0x00017] in <6511abc92ea846e0b34f20222561029d>:0
  at Microsoft.Dafny.Translator+ExpressionTranslator.TrExpr (Microsoft.Dafny.Expression expr) [0x01a54] in <6511abc92ea846e0b34f20222561029d>:0
  at Microsoft.Dafny.Translator.FunctionAxiom (Microsoft.Dafny.Function f, Microsoft.Dafny.Expression body, System.Collections.Generic.ICollection`1[T] lits, Microsoft.Dafny.TopLevelDecl overridingClass) [0x00ab1] in <6511abc92ea846e0b34f20222561029d>:0
  at Microsoft.Dafny.Translator.AddFunctionAxiom (Microsoft.Dafny.Function f, Microsoft.Dafny.Expression body) [0x00001] in <6511abc92ea846e0b34f20222561029d>:0
  at Microsoft.Dafny.Translator.AddClassMember_Function (Microsoft.Dafny.Function f) [0x0008d] in <6511abc92ea846e0b34f20222561029d>:0
  at Microsoft.Dafny.Translator.AddFunction_Top (Microsoft.Dafny.Function f) [0x00020] in <6511abc92ea846e0b34f20222561029d>:0
  at Microsoft.Dafny.Translator.AddClassMembers (Microsoft.Dafny.ClassDecl c, System.Boolean includeMethods) [0x0034a] in <6511abc92ea846e0b34f20222561029d>:0
  at Microsoft.Dafny.Translator.DoTranslation (Microsoft.Dafny.Program p, Microsoft.Dafny.ModuleDefinition forModule) [0x00457] in <6511abc92ea846e0b34f20222561029d>:0
  at Microsoft.Dafny.Translator+<Translate>d__47.MoveNext () [0x000ea] in <6511abc92ea846e0b34f20222561029d>:0
  at Microsoft.Dafny.DafnyDriver+<Translate>d__6.MoveNext () [0x00119] in <1726fb2ac8b046fb9f3fa158ff53a56f>:0
  at Microsoft.Dafny.DafnyDriver.Boogie (System.String baseName, System.Collections.Generic.IEnumerable`1[T] boogiePrograms, System.String programId, System.Collections.Generic.Dictionary`2[System.String,Microsoft.Boogie.PipelineStatistics]& statss, Microsoft.Boogie.PipelineOutcome& oc) [0x0012f] in <1726fb2ac8b046fb9f3fa158ff53a56f>:0
  at Microsoft.Dafny.DafnyDriver.ProcessFiles (System.Collections.Generic.IList`1[T] dafnyFiles, System.Collections.ObjectModel.ReadOnlyCollection`1[T] otherFileNames, Microsoft.Dafny.ErrorReporter reporter, System.Boolean lookForSnapshots, System.String programId) [0x00222] in <1726fb2ac8b046fb9f3fa158ff53a56f>:0
  at Microsoft.Dafny.DafnyDriver.ThreadMain (System.String[] args) [0x0003a] in <1726fb2ac8b046fb9f3fa158ff53a56f>:0
  at Microsoft.Dafny.DafnyDriver+<>c__DisplayClass1_0.<Main>b__0 () [0x00001] in <1726fb2ac8b046fb9f3fa158ff53a56f>:0
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) [0x00014] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x0002b] in <f7cedb26ce694cd281e4450870c6fe49>:0
  at System.Threading.ThreadHelper.ThreadStart () [0x00008] in <f7cedb26ce694cd281e4450870c6fe49>:0
```

--- a/Source/Dafny/Translator.cs
+++ b/Source/Dafny/Translator.cs
@@ -17354,7 +17354,7 @@ namespace Microsoft.Dafny {
         if (call != null && call.Function != null && call.Function.ReadsHeap) {
           foundHeap = true;
         }
-        if (expr is ApplyExpr) {
+        if (expr is ApplyExpr || expr is SeqConstructionExpr) {
           foundHeap = true;
         }
         ThisExpr thisExpr = expr as ThisExpr;

--- a/Test/git-issues/git-issue-334.dfy
+++ b/Test/git-issues/git-issue-334.dfy
@@ -1,0 +1,12 @@
+// RUN: %dafny "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+function method FromAtoB(a: seq<int>): seq<int> {
+  // regression: the following had once crashed the translator
+  seq(|a|, i => a[i])  // error: index out of bounds
+}
+
+function method FromAtoC(a: seq<int>): seq<int> {
+  // regression: the following had once crashed the translator
+  seq(|a|, i requires 0 <= i < |a| => a[i])
+}

--- a/Test/git-issues/git-issue-334.dfy.expect
+++ b/Test/git-issues/git-issue-334.dfy.expect
@@ -1,0 +1,7 @@
+git-issue-334.dfy(6,17): Error: index out of range
+Execution trace:
+    (0,0): anon0
+    (0,0): anon6_Else
+    (0,0): anon7_Then
+
+Dafny program verifier finished with 1 verified, 1 error

